.github/workflows/hash-backend-cd.yml

```diff
@@ -0,0 +1,139 @@
+          context: ${{ github.workspace }}
+          push: true
+          file: ${{ github.workspace }}/packages/hash/docker/api/prod/Dockerfile
+          tags: ${{ env.AWS_ECR_URL }}/${{ env.HASH_API_IMAGE_NAME }}:latest
+
+  build-realtime:
+    name: Build and push HASH realtime image
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: ./.github/actions/docker-ecr-login
+        with:
+          AWS_ACCESS_KEY_ID: ${{ env.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ env.AWS_SECRET_ACCESS_KEY }}
+          AWS_REGION: ${{ env.AWS_REGION }}
+
+      - name: Build and push realtime
+        uses: docker/build-push-action@v2
+        env:
+          HASH_REALTIME_IMAGE_NAME: ${{ env.HASH_REALTIME_RESOURCE_NAME }}ecr
+        with:
+          context: ${{ github.workspace }}
+          push: true
+          file: ${{ github.workspace }}/packages/hash/docker/realtime/prod/Dockerfile
+          tags: ${{ env.AWS_ECR_URL }}/${{ env.HASH_REALTIME_IMAGE_NAME }}:latest
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache-new
+
+  build-searchloader:
+    name: Build and push HASH searchloader image
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: ./.github/actions/docker-ecr-login
+        with:
+          AWS_ACCESS_KEY_ID: ${{ env.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ env.AWS_SECRET_ACCESS_KEY }}
+          AWS_REGION: ${{ env.AWS_REGION }}
+
+      - name: Build and push search-loader
+        uses: docker/build-push-action@v2
+        env:
+          HASH_SEARCHLOADER_IMAGE_NAME: ${{ env.HASH_SEARCHLOADER_RESOURCE_NAME }}ecr
+        with:
+          context: ${{ github.workspace }}
+          push: true
+          file: ${{ github.workspace }}/packages/hash/docker/search-loader/prod/Dockerfile
+          tags: ${{ env.AWS_ECR_URL }}/${{ env.HASH_SEARCHLOADER_IMAGE_NAME }}:latest
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache-new
+
+  deploy:
+    name: Deploy HASH images
+    runs-on: ubuntu-latest
+    needs:
+      - build-api
+      - build-realtime
+      - build-searchloader
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: ./.github/actions/docker-ecr-login
+        with:
+          AWS_ACCESS_KEY_ID: ${{ env.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ env.AWS_SECRET_ACCESS_KEY }}
+          AWS_REGION: ${{ env.AWS_REGION }}
+
+      - name: Prepare dev tools for DB migration
+        run: yarn install
+        shell: bash
+
+      # TODO: should we stop the services before migrating the db?
+      - name: Migrate database
+        run: |
+          HASH_MIGRATION_PG_URL=postgres://${{ env.HASH_POSTGRES_MIGRATE_USER_PASS }}@${{ env.HASH_POSTGRES_HOST_DB }} \
+            yarn workspace @hashintel/hash-datastore pg:migrate up
```

Comment:
I think I know the answer, but just want to double-check: if the migrate database step fails the deployment action is aborted, and the images are not updated, correct?

Reply:
The images would still be uploaded, but they wouldn't be redeployed, as this step would stop the next ones.
It's a good question as we could get into some weird state. I don't know if we can handle all of the steps atomically, though

Reply:
My understanding is that re-running the flow should be possible, which would let us try the migration again if the think the error was circumstantial. If the error persists we could then try to address the error in the branch.

Perhaps down the line having a dry run of the migrations as a CI check would be a good idea, as it'd let us catch any migration errors in advance.

Reply:
Dry-runs would be great. I'm wondering if it would then be beneficial for us to add a manual dispatch for this pipeline as well? So we don't need to force push to main when we want to retrigger this deployment pipeline